[UI-side compositing] Cannot grab the scrubber to scroll to the end of a long page (build log)

nmoucht opened the pull request:
#### 0f2ddd2e32274c34d687b6263b7664cc4c9aa5b2
<pre>
[UI-side compositing] Cannot grab the scrubber to scroll to the end of a long page (build log)
<a href="https://bugs.webkit.org/show_bug.cgi?id=259167">https://bugs.webkit.org/show_bug.cgi?id=259167</a>
rdar://112154989

Reviewed by Simon Fraser.

After <a href="https://commits.webkit.org/265731@main">https://commits.webkit.org/265731@main</a> removed NSScrollerImps in the web process, when
ScrollbarThemeMac::minimumThumbLength would query the scrollbar map, it wouldn&apos;t be able to
find the NSScrollerImp, so it would return 0. This resulted in users not being able to drag
the scrollbar. To resolve this, plumb the knobMinLength property to the web process, where it
is tracked by RemoteScrollbarsController.

* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::requestScrollToPosition):
(WebCore::AsyncScrollingCoordinator::scrollingTreeNodeScrollbarMinimumThumbLengthDidChange):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingTree.h:
(WebCore::ScrollingTree::scrollingTreeNodeScrollbarMinimumThumbLengthDidChange):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::scrollbarMinimumThumbLengthDidChange):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(WebCore::ScrollerMac::attach):
(WebCore::ScrollerMac::updateScrollbarStyle):
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::minimumThumbLength const):
* Source/WebCore/platform/Scrollbar.h:
* Source/WebCore/platform/ScrollbarsController.h:
(WebCore::ScrollbarsController::setScrollbarMinimumThumbLength):
(WebCore::ScrollbarsController::minimumThumbLength):
* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::ScrollbarThemeMac::hasThumb):
(WebCore::ScrollbarThemeMac::minimumThumbLength):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeScrollbarMinimumThumbLengthDidChange):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeScrollbarMinimumThumbLengthDidChange):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
(WebKit::RemoteScrollbarsController::setScrollbarMinimumThumbLength):
(WebKit::RemoteScrollbarsController::minimumThumbLength):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.messages.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
(WebKit::RemoteScrollingCoordinator::scrollingTreeNodeScrollbarMinimumThumbLengthDidChange):

Canonical link: <a href="https://commits.webkit.org/266124@main">https://commits.webkit.org/266124@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/479d1fae3a6b70717141033c57e1a1d1ae770704

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12695 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13021 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13347 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14434 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12143 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12756 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15525 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13039 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14846 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12859 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13616 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10745 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14878 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10898 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11493 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18570 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11978 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11660 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14855 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12126 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10041 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11379 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15703 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1470 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11978 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerMac.h:
```diff
@@ -70,8 +70,10 @@ class ScrollerMac {
+    void updateMinimumKnobLength(int);
     void detach();
 private:
+    int m_minimumKnobLength;
```
Give it an initializer: `{ 0 }`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -332,6 +332,7 @@ - (void)invalidate
     m_scrollerImpDelegate = adoptNS([[WebScrollerImpDelegateMac alloc] initWithScroller:this]);
     m_scrollerImp = [NSScrollerImp scrollerImpWithStyle:nsScrollerStyle(m_pair.scrollbarStyle()) controlSize:nsControlSizeFromScrollbarWidth(m_pair.scrollbarWidthStyle()) horizontal:m_orientation == ScrollbarOrientation::Horizontal replacingScrollerImp:nil];
```
If we called `setScrollerImp()` here that would take care of the `updateMinimumKnobLength() call.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -370,6 +371,7 @@ - (void)invalidate
 void ScrollerMac::updateScrollbarStyle()
 {
     m_scrollerImp = [NSScrollerImp scrollerImpWithStyle:nsScrollerStyle(m_pair.scrollbarStyle()) controlSize:nsControlSizeFromScrollbarWidth(m_pair.scrollbarWidthStyle()) horizontal:m_orientation == ScrollbarOrientation::Horizontal replacingScrollerImp:takeScrollerImp().get()];
```
If we called `setScrollerImp()` here that would take care of the `updateMinimumKnobLength() call.

Not sure how we get away with not setting the delegate like we did above?

nmoucht replied:
Fixed. The delegate gets passed over through `replacingScrollerImp`.

smfr commented on Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
```diff
@@ -438,9 +438,12 @@ static IntRect buttonRepaintRect(const IntRect& buttonRect, ScrollbarOrientation
-    BEGIN_BLOCK_OBJC_EXCEPTIONS
-    return [scrollbarMap().get(&scrollbar) knobMinLength];
-    END_BLOCK_OBJC_EXCEPTIONS
+    if (scrollbar.shouldRegisterScrollbar()) {
```
This stuff makes me wonder why we don't just ask the scrollbar for its minimumKnobLength. Why is the map lookup needed? Then we could have a polymorphic Scrollbar class where there's a subclass that never has an NSScrollerImp.

nmoucht replied:
Do you mean rather than having the map, have the Scrollbar class hold the NSScrollerImp if UI-side compositing is off, and have a Scrollbar not held if UI-side compositing is on? If yes then yeah I think that would be a good cleanup for this area once this lands.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:
```diff
@@ -49,13 +49,18 @@ class RemoteScrollbarsController final : public WebCore::ScrollbarsController {
     bool m_horizontalOverlayScrollbarIsVisible { false };
     bool m_verticalOverlayScrollbarIsVisible { false };
+
+    int m_horizontalMinimumThumbLength;
```
Initialize these.

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.h:
```diff
@@ -97,6 +97,7 @@ class RemoteScrollingCoordinator final : public WebCore::AsyncScrollingCoordinat
     void startDeferringScrollingTestCompletionForNode(WebCore::ScrollingNodeID, WebCore::WheelEventTestMonitor::DeferReason);
     void stopDeferringScrollingTestCompletionForNode(WebCore::ScrollingNodeID, WebCore::WheelEventTestMonitor::DeferReason);
     void scrollingTreeNodeScrollbarVisibilityDidChange(WebCore::ScrollingNodeID, WebCore::ScrollbarOrientation, bool);
+    void scrollingTreeNodeScrollbarMinimumThumbLengthDidChange(WebCore::ScrollingNodeID nodeID, WebCore::ScrollbarOrientation orientation, int minimumThumbLength);
```
The base class has a function with the same signature; not sure why this and the function above are not virtual on the base class.

nmoucht replied:
Fixed to just implement this on RemoteScrollingCoordinator.

nmoucht committed:
[UI-side compositing] Cannot grab the scrubber to scroll to the end of a long page (build log)
https://bugs.webkit.org/show_bug.cgi?id=259167
rdar://112154989

Reviewed by Simon Fraser.

After https://commits.webkit.org/265731@main removed NSScrollerImps in the web process, when
ScrollbarThemeMac::minimumThumbLength would query the scrollbar map, it wouldn't be able to
find the NSScrollerImp, so it would return 0. This resulted in users not being able to drag
the scrollbar. To resolve this, plumb the knobMinLength property to the web process, where it
is tracked by RemoteScrollbarsController.

* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::requestScrollToPosition):
(WebCore::AsyncScrollingCoordinator::scrollingTreeNodeScrollbarMinimumThumbLengthDidChange):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingTree.h:
(WebCore::ScrollingTree::scrollingTreeNodeScrollbarMinimumThumbLengthDidChange):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::scrollbarMinimumThumbLengthDidChange):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(WebCore::ScrollerMac::attach):
(WebCore::ScrollerMac::updateScrollbarStyle):
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::minimumThumbLength const):
* Source/WebCore/platform/Scrollbar.h:
* Source/WebCore/platform/ScrollbarsController.h:
(WebCore::ScrollbarsController::setScrollbarMinimumThumbLength):
(WebCore::ScrollbarsController::minimumThumbLength):
* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::ScrollbarThemeMac::hasThumb):
(WebCore::ScrollbarThemeMac::minimumThumbLength):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeScrollbarMinimumThumbLengthDidChange):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeScrollbarMinimumThumbLengthDidChange):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
(WebKit::RemoteScrollbarsController::setScrollbarMinimumThumbLength):
(WebKit::RemoteScrollbarsController::minimumThumbLength):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.messages.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
(WebKit::RemoteScrollingCoordinator::scrollingTreeNodeScrollbarMinimumThumbLengthDidChange):

Canonical link: https://commits.webkit.org/266124@main
